package NYT.NObjectClient.NProto;

import "yt/core/misc/proto/guid.proto";

////////////////////////////////////////////////////////////////////////////////

message TReqExecute
{
    // For forwarded and redirected requests, contains the original request id.
    optional NYT.NProto.TGuid original_request_id = 6;

    // Each Execute call carries a number of YPath requests.
    // Each of these requests is represented by a message.
    // Each message, in turn, is a sequence of parts.
    // This array contains the lengths of those part sequences.
    // In particular, its length is just the number of encapsulated YPath requests.
    repeated int32 part_counts = 1;

    // If |true|, the master responding to the query will make no attempt to synchronize
    // its state with the upstream servers.
    optional bool suppress_upstream_sync = 2 [default = false];

    // If |true|, the master is allowed to send a partial response to this request.
    optional bool allow_backoff = 3 [default = false];

    // If |true|, the client supports TRspExecute.subresponses and TRspExecute.uncertain_subrequest_indexes.
    optional bool supports_portals = 5;
}

message TRspExecute
{
    // COMPAT(babenko)
    // Similar to TReqExecute::part_counts, but denotes the number of parts in each
    // encapsulated response.
    // NB: a batched response may have less subresponses than its corresponding request
    // does subrequests if TReqExecute::allow_backoff was |true| (because of timeouts).
    repeated int32 part_counts = 1;

    // Hydra revision at the beginning of each YPath request execution.
    // COMPAT(babenko)
    repeated uint64 revisions = 2;

    message TSubresponse
    {
        // The index of the subrequest.
        required int32 index = 1;
        // The number of parts in this subresponse.
        required int32 part_count = 2;
        // Hydra revision at the beginning of the subrequest execution.
        required uint64 revision = 3;
    }
    // If non-empty, gives the detailed subresponse information in the new format.
    repeated TSubresponse subresponses = 3;

    // The indexes of subrequests that were forwarded to another master but
    // for which no response was received; the client must annotate the subsequent attempts with
    // "retry" flag.
    repeated int32 uncertain_subrequest_indexes = 4;

    // If true, two-level cache is enabled at master.
    optional bool two_level_cache_enabled = 5 [default = false];
}

////////////////////////////////////////////////////////////////////////////////

message TReqGCCollect
{ }

message TRspGCCollect
{ }

////////////////////////////////////////////////////////////////////////////////
